Feature: 로그인 & 로그아웃 msw 적용

sysysysyb opened the pull request:
## 🚀 PR 요약

로그인 테스트 페이지에 로그인 & 로그아웃 msw 로직을 적용했습니다.

## ✏️ 변경 유형

- [x] feat: 새로운 기능 추가

## ✅ 체크리스트

- [x] 커밋 컨벤션에 맞춰 커밋 메시지를 작성했습니다.
- [x] 커밋에 관련된 이슈 번호를 포함했습니다.
- [x] 실행에 문제가 없는지 테스트했습니다.

## 🗒️ 상세 내용 (선택)

### 작업 사항

- **useLogin**, **useLogout** hook 추가
- **useUserInformation** 훅 수정
  - 로그인 시에만 사용자 정보를 불러오도록 **enabled** 옵션 추가
- 사용자 정보 조회(`useUserInformation` 훅) api 요청 후 상태 코드 200을 반환하면 로그인을 한 상태이고, 에러를 반환하면 로그인을 하지 않은 상태라고 판단
- 액세스 토큰 재발급 api에서 401 에러 반환 시 axois interceptor
  - api 요청 후 401 에러 반환 시 액세스 토큰 재발급 api 요청
  - 액세스 토큰 재발급 api에서 401 에러 반환 시, 라우트 경로에 **my-page**가 들어가면 마이페이지에 접근하지 못하도록 로그인 페이지로 강제 이동

### 참고 사항

- 실제 로그인 페이지에는 따로 이슈를 생성해서 헤더 UI 변경 부분까지 한번에 추가할 예정입니다.
- 테스트 페이지의 입력 필드는 이메일, 비밀번호가 틀렸을 때 에러 발생하는 것을 확인하기 위해 유효성 검증 없이 간단하게 추가했습니다. 
  기본 설정된 값을 바꾸지 않고 로그인하면 문제 없이 로그인  되고, 값을 바꾸고 로그인하면 에러가 발생하여 로그인 되지 않습니다.
- 소셜 로그인 버튼은 이후 서버 연결할 때 이 곳에서 테스트하기 위해 버튼 UI만 추가해놓은 것으로 지금은 클릭해도 실행되지 않습니다.
- (250918 20:25 추가) 사용자 정보 조회 api의 반환 결과에 따라 로그인 여부를 판단합니다. 실제 서버 연결 시 쿠키에 저장되는 토큰들은 HttpOnly가 설정되어 확인이 불가능하지만, msw 로직을 테스트하기 위해 임시로 msw 로그인 시 HttpOnly가 설정되지 않은 **access-token**과 **refresh-token**이 추가되도록 하였습니다.
- (250918 20:25 추가) 트러블슈팅: [로그인 & 로그아웃 여부 관리하기](https://www.notion.so/272caf5650aa806db5e5d224227604ef?source=copy_link)

### 스크린 샷

**msw 적용 테스트를 위해 임시로 쿠키에서 액세스 토큰, 리프레시 토큰 관리**
![StudyHub-Chrome2025-09-1819-13-31-ezgif com-video-to-gif-converter](https://github.com/user-attachments/assets/5c16ce5c-ae37-4510-8714-7554ac38c36a)

**이메일 또는 비밀번호 오류 시 에러 발생**
![StudyHub-Chrome2025-09-1815-15-52-ezgif com-video-to-gif-converter](https://github.com/user-attachments/assets/69d9e808-8562-4fdb-8ddc-664f4a15cd7d)

## 🔗 연관된 이슈

> closes #122

sysysysyb committed:
✨ Feature: withCredentials 추가 및 refresh 로직 임시 추가 (#122)

sysysysyb committed:
✨ Feature: useLogin 훅 추가 (#122)

sysysysyb committed:
♻️ Refactor: queryKey 수정 (#122)

sysysysyb committed:
✨ Feature: 베럴 파일에 추가 및 수정 (#122)

sysysysyb committed:
✨ Types: 로그인 관련 타입 추가 (#122)

sysysysyb committed:
✨ Feature: 로그인 mockdata 추가 (#122)

sysysysyb committed:
♻️ Refactor: 파일 이름 수정 및 핸들러 추가 (#122)

sysysysyb committed:
✨ Feature: 테스트 페이지에 로그인 로직 추가 (#122)

sysysysyb committed:
✨ Feature: useLogout 훅 추가 (#122)

로그아웃 api 확정되면 수정 예정

sysysysyb committed:
✨ Feature: 테스트 페이지에 로그아웃 추가 및 훅 option 수정 (#122)

sysysysyb committed:
✨ Feature: 세션 스토리지에서 로그인 여부 관리하는 유틸 추가 (#122)

세션 스토리지로 로그인 여부 확인해서 true 또는 false 반환하는 함수, 세션 스토리지에 로그인 성공 상태를 저장하는 함수, (로그아웃 시) 세션 스토리지에서 로그인 성공 상태를 삭제하는 함수

sysysysyb committed:
✨ Feature: 세션 스토리지 로그인 여부 관리 유틸 배럴 파일에 추가 (#122)

sysysysyb committed:
♻️ Refactor: useLogin 훅 수정 (#122)

api 명세서에 맞게 쿼리키 수정 / 로그인 성공 시 세션 스토리지에 로그인 성공 상태 저장하는 함수 적용 / 에러 발생 시 토스트 메시지 적용

sysysysyb committed:
♻️ Refactor: useLogout 훅 추가 (#122)

api 명세서에 따라 쿼리키 수정 / cancelQueries 삭제 / 세션 스토리지에서 로그인 성공 상태 삭제하는 함수 적용

sysysysyb committed:
♻️ Refactor: useUserInformation 수정 (#122)

세션 스토리지에서 로그인 성공 상태가 확인되면 실행되도록 enabled 옵션 추가

sysysysyb committed:
♻️ Refactor: auth 관련 msw 핸들러 수정 (#122)

api 명세서에 따라 엔드포인트 수정 / 로그인 에러 발생 로직 추가

sysysysyb committed:
♻️ Refactor: 로그인 테스트 페이지 수정 (#122)

사용자 정보 가져오기 테스트 페이지를 로그인 테스트 페이지로 변경 / 사용자 정보 가져오기 요청 enabled 옵션을 훅에서 관리하도록 수정 / 이후 소셜 로그인 테스트를 위해 버튼 UI만 임시 추가

sysysysyb committed:
✨ Feature: 이메일, 비밀번호 입력 필드 추가 (#122)

이메일 또는 비밀번호가 틀릴 경우 토스트 메시지가 출력되며 로그인이 실행되지 않는 것을 보여주기 위해 추가

sysysysyb committed:
♻️ Refactor: 소셜 로그인 버튼 수정 (#122)

소셜 로그인 버튼 컴포넌트에 props로 kakao 또는 naver을 전달하여 버튼 종류 선택

sysysysyb committed:
✨ Feature: 수정된 소셜 로그인 버튼 적용 (#122)

sysysysyb committed:
💡 Comment: api 토큰 리프레시 interceptor 로직 주석 삭제 (#122)

주석으로 임시 추가 후 서버 연결 시 약간의 수정을 거쳐 사용하려 했는데 api 명세서를 보니 엔드포인트 등 전체적으로 수정 중인 것 같아 삭제하고 이후 다시 추가 예정

JaeHyunLee123 commented on src/types/api-response-types/auth.ts:
```diff
@@ -0,0 +1,15 @@
+export interface UserLogin {
+  email: string
+  password: string
+}
```
해당 타입은 mutation 시 클라이언트가 서버에게 보내는 타입입니다. 그래서 api-response-types 폴더 안에 있는게 조금 어색할 수도 있을 것 같습니다.

별도로 api-request-types 폴더를 만드는 건 어떨까요?

제가 api-request-types.ts 파일에 임시로 정의해둔 타입이 있는데 폴더 따로 만드시면 거기에 통합하겠습니다.

sysysysyb replied:
**api-request-types** 폴더 추가하고 `UserLogin` 타입 해당 폴더로 이동했습니다!!

JaeHyunLee123 commented on src/tests/LoginTest.tsx:
```diff
@@ -0,0 +1,88 @@
+    const dd = date?.substring(8, 10)
+
+    return `${yyyy}. ${mm}. ${dd}.`
+  }
```
utils에 승원님이 정의해두신 유틸 함수 중에 비슷한 역할을 하는게 있습니다. 추후에 리팩토링 진행해서 통합 고려해봐도 좋을 것 같습니다.

sysysysyb replied:
**formatted-dates.ts** 파일에서 관련 함수 관리하는 게 맞는 것 같아서 `FormattedDateWithDots`로 함수명 변경하고 이동했습니다!!

dirage1 reviewed (dismissed):
수고하셨습니다! getIsLoggedin으로 추후에 마이페이지 관련 hook들도 수정하겠습니다

JaeHyunLee123 commented on src/utils/session-auth.ts:
세션 스토리지에 저장 시 유저가 탭을 닫았다가 다시 열면 자동으로 스토리지에 있던 값들이 사라져 로그아웃 됩니다. 의도한 사항이면 이대로 둬도 좋고 탭을 열었다가 닫아도 로그인을 유지 시키고 싶으면 로컬 스토리지 사용 고려해봐도 좋을 것 같습니다.

sysysysyb replied:
아 이 부분은 저도 인지하고 있는데 일단 탭을 닫으면 쿠키도 사라질 것이라고 가정하고 이렇게 작성했습니다! 근데 아무래도 확실히 확인하고 넘어가는 게 맞을 것 같아서 조교님 코치님과 논의 후에 확정하도록 하겠습니다!

탭을 닫았을 때 쿠키에 담은 토큰이 초기화되면 세션 스토리지에, 남아있으면 로컬 스토리지에서 관리하는 방향으로 하겠습니다!!

sysysysyb replied:
사용자 정보 조회 api 요청을 보내고 반환된 응답의 상태 코드가 `200`이면 **로그인한 상태**, `401` 등 에러가 발생하면 **로그인 하지 않은 상태**로 판단하도록 수정했습니다!!

JaeHyunLee123 reviewed (changes requested):
수고하셨습니다! 몇가지 논의할 사항이 있어서 해당 부분에 리뷰 남겼습니다.

sysysysyb committed:
🚚 Rename: auth 관련 response 타입 파일명 수정 (#122)

sysysysyb committed:
✨ Feature: auth 관련 request 타입 파일 추가 (#122)

sysysysyb committed:
✨ Feature: 날짜 변환 함수 유틸 파일에 옮긴 후 적용 (#122)

sysysysyb committed:
🔥 Delete: 세션 스토리지 로그인 여부 관리 유틸 삭제 (#122)

sysysysyb committed:
♻️ Refactor: useUserInformation 수정 (#122)

세션 스토리지 관련 유틸 삭제 / 에러 발생해도 한번만 실행하도록 retry 옵션 추가

sysysysyb committed:
♻️ Refactor: auth 관련 msw 핸들러 수정 (#122)

msw 적용을 위해 쿠키에 임시로 담을 access-token과 refresh-token 추가 / 로그인 시 쿠키에 두 토큰이 담기도록 수정 / 로그아웃 시 두 토큰이 삭제되도록 수정 / 사용자 정보를 가져올 때 액세스 토큰이 존재하는 지 체크 / 액세스 토큰 재발급 msw 핸들러 추가

sysysysyb committed:
✨ Feature: 토큰 재발급 refresh 관련 interceptor 추가 (#122)

401 에러 발생 시 refresh 요청 / 만약 refresh 요청에서 401 에러 발생 시 현재 페이지가 마이페이지일 경우 로그인 페이지로 이동하고 그 외 경우는 무시

sysysysyb committed:
♻️ Refactor: Refactor: 사용자 정보 가져오기 에러 여부로 렌더링 관리 (#122)

사용자 정보 가져오기 api에서 에러 발생 시 로그인 버튼 렌더링 / 200 반환 시 로그아웃 및 받아온 사용자 정보 렌더링

JaeHyunLee123 commented on src/utils/formatted-dates.ts:
이 파일에 있는 함수들 다 첫 글자 소문자로 바꾸면 좋을 것 같습니다. 별도 이슈 파서 진행하겠습니다. (#166)

sysysysyb replied:
넵 감사합니다!!!

JaeHyunLee123 reviewed (approved):
트러블 슈팅 읽어보니 흐름을 직관적으로 이해할 수 있었습니다. 개발 편의성과 유저 경험을 모두 고려한 내용이 인상적입니다. 수고하셨습니다!

dirage1 reviewed (approved):
수고하셨습니다
